test/status_file.c

```diff
@@ -360,6 +369,131 @@ boot-id=e02a2afe-cf45-4d50-a3f3-c223ca0f480a\n\
+
+static void status_file_test_datadir(StatusFileFixture *fixture, gconstpointer user_data)
+{
+	GHashTable *slots = r_context()->config->slots;
+	GError *ierror = NULL;
+	gboolean res;
+
+	g_assert_nonnull(r_context()->config->statusfile_path);
+
+	g_assert_nonnull(slots);
+	RaucSlot *slot = g_hash_table_lookup(slots, "rootfs.0");
+	g_assert_nonnull(slot);
+
+	/* Create status */
+	if (slot->status)
+		r_slot_free_status(slot->status);
+	slot->status = g_new0(RaucSlotStatus, 1);
+	slot->status->status = g_strdup("ok");
+	slot->status->checksum.type = G_CHECKSUM_SHA256;
+	slot->status->checksum.digest = g_strdup(DIGEST_INITIAL);
+	slot->status->checksum.size = 512;
+
+	g_autofree gchar *initial_checksum_dir = r_slot_get_checksum_data_directory(slot, NULL, &ierror);
+	g_assert_nonnull(initial_checksum_dir);
+	g_assert_no_error(ierror);
+
+	g_autofree gchar *initial_file = write_tmp_file(initial_checksum_dir, "testfile", "content", &ierror);
+	g_assert_nonnull(initial_file);
+	g_assert_no_error(ierror);
+	g_assert_true(g_file_test(initial_file, G_FILE_TEST_IS_REGULAR));
+
+	GStatBuf initial_stat = {0};
+	g_assert_cmpint(g_lstat(initial_file, &initial_stat), ==, 0);
+
+	RaucChecksum other_checksum = {
+		.type = G_CHECKSUM_SHA256,
+		.digest = g_strdup(DIGEST_OTHER),
+	};
+	g_autofree gchar *other_checksum_dir = r_slot_get_checksum_data_directory(slot, &other_checksum, &ierror);
+	g_assert_nonnull(other_checksum_dir);
+	g_assert_no_error(ierror);
+	g_clear_pointer(&other_checksum.digest, g_free);
+
+	g_autofree gchar *other_file = write_tmp_file(other_checksum_dir, "testfile", "other-content", &ierror);
+	g_assert_nonnull(other_file);
+	g_assert_no_error(ierror);
+	g_assert_true(g_file_test(other_file, G_FILE_TEST_IS_REGULAR));
+
+	res = r_slot_status_save(slot, &ierror);
+	g_assert_no_error(ierror);
+	g_assert_true(res);
+	g_assert_true(g_file_test(initial_file, G_FILE_TEST_IS_REGULAR));
+	g_assert_false(g_file_test(other_file, G_FILE_TEST_IS_REGULAR));
+
+	/* Simulate starting an installation */
+	g_assert_nonnull(slot->status);
+	r_replace_strdup(&slot->status->status, "pending");
+	g_clear_pointer(&slot->status->checksum.digest, g_free);
+	slot->status->checksum.size = 0;
+
+	res = r_slot_move_checksum_data_directory(slot, DIGEST_INITIAL, "unknown", &ierror);
+	g_assert_no_error(ierror);
+	g_assert_true(res);
+
+	g_autofree gchar *unknown_checksum_dir = r_slot_get_checksum_data_directory(slot, NULL, &ierror);
+	g_assert_nonnull(unknown_checksum_dir);
+	g_assert_no_error(ierror);
+	g_clear_pointer(&other_checksum.digest, g_free);
+
+	g_autofree gchar *unknown_file = g_build_filename(unknown_checksum_dir, "testfile", NULL);
+	GStatBuf unknown_stat = {0};
+	g_assert_cmpint(g_lstat(unknown_file, &unknown_stat), ==, 0);
+	g_assert_cmpuint(initial_stat.st_ino, ==, unknown_stat.st_ino);
+
+	g_assert_false(g_file_test(initial_file, G_FILE_TEST_IS_REGULAR));
+	g_assert_false(g_file_test(other_file, G_FILE_TEST_IS_REGULAR));
```

Comment:
Does it make sense to continue testing that `other_file` does not exist?
I don't see a way how it could accidentally be re-generated.

Reply:
Better safe than sorry, given how much confusion we already had in this area.